### PR TITLE
[algo] Add On Policy Self Distiation with e2e recipe

### PR DESCRIPTION
**Pull request by gxlvera**
# Algo (OPD): On-Policy *Self*-Distillation (OPSD)

## Summary

This PR adds **On-Policy Self-Distillation (OPSD)** ([arXiv:2601.18734](https://arxiv.org/abs/2601.18734)) on top of verl's existing On-Policy Distillation (OPD) / `teacher_loop` infrastructure.

OPD trains a student on its own rollouts while a **separate, larger, frozen teacher** provides dense token-level supervision — but it requires a second model on a dedicated teacher pool. OPSD removes the *external* teacher: the **same architecture/checkpoint** plays both roles, differing only in **conditioning context**. The teacher conditions on the question `x` *plus* privileged information `y*` (the verified reference solution / CoT); the student conditions on `x` only. Gradients flow through the student only.

Two teacher-weight sources are supported via config:
- **`frozen` (default, reproduces the paper §4.1):** teacher = the initial checkpoint `θ₀`, served by the existing OPD teacher pool (point the teacher `model_path` at the student base).
- **`live` (Algorithm-1 variant):** teacher = current student weights served by the rollout engine, no separate teacher GPUs. Config surface + validation are in place; the `frozen` path is fully wired end-to-end in this cut.

The whole thing reuses OPD's loss layer, padding, and engine forward **verbatim** — only (S1) the input sequence and (S2) the output layout differ. See `docs/algo/opsd.md` for the full design.

### Core OPSD wiring
- **`verl/experimental/teacher_loop/teacher_manager.py`**
  - `build_privileged_sequence` (S1): rebuild the teacher prompt as a *proper chat turn* from `(problem x, reference y*)` via the chat template (matching the reference impl), truncating `y*` to `max_reference_length`; returns `[teacher_prompt, ŷ]` + prefix length.
  - `remap_privileged_to_student_layout` (S2): remap the privileged teacher tensor back onto the student `[x, ŷ]` layout so `_pad_teacher_outputs` and downstream causal alignment stay byte-for-byte identical to OPD (works for both 1-D and 2-D top-k tensors).
  - `compute_self_distill_logprobs_single`: reuses `compute_teacher_logprobs_single` for the forward, wrapping it with S1/S2.
- **`verl/experimental/agent_loop/agent_loop.py`**: route teacher log-probs through the privileged builder when OPSD is enabled; pass the tokenizer into the teacher manager; `NotImplementedError` guard for `teacher_weights='live'` follow-up.

### Config
- **`verl/workers/config/distillation.py` + `distillation.yaml`**: new `SelfDistillationConfig` (`enabled`, `teacher_weights`, `reference_key`, `problem_key`, `teacher_template`, `teacher_thinking`, `max_reference_length`, `max_bridge_length`). Teacher context sizing extended by `privileged_extra = max_reference_length + max_bridge_length` in `validate_and_prepare_for_distillation` (defaults to `0`, so plain OPD arithmetic is unchanged).

### Loss (reference parity, FSDP `forward_kl_topk`)
- **`verl/trainer/distillation/fsdp/losses.py` + `distillation/losses.py`**:
  - `clip_tau`: per-vocab pointwise KL clipping *before* the vocab sum (distinct from the existing post-sum `loss_max_clamp`).
  - `loss_temperature`: softmax temperature applied to **both** student and teacher logits over the top-k support.
  - `clamp_negative_to_zero` (default `True` = original verl behavior): gate the `clamp_min(0.0)` floor so the reference's negative-allowed per-token loss can be reproduced.
  - **Sampling-temperature undo**: multiply back the engine's rollout-temperature division so distillation targets are built from **raw** logits (avoids dividing the student by temperature twice).
- **Megatron path**: `clip_tau` and `student_logits_temperature` raise `NotImplementedError` (FSDP-only in this cut; the custom backward must zero clipped-entry gradients).

### Engine / training
- **`verl/workers/engine/fsdp/transformer_impl.py`**: pass `student_logits_temperature` into the logits processor (both padded and rmpad paths).
- **`verl/workers/config/optimizer.py`**: allow `lr_scheduler_type='linear'`; FSDP engine wires HF's `get_linear_schedule_with_warmup` for reference parity.
- **`verl/trainer/ppo/ray_trainer.py`**: mirror core metrics into a `train/` wandb section (`train/loss`, `train/learning_rate`, `train/grad_norm`) using HF-Trainer metric names so verl and reference runs overlay directly. Aliases only; original keys untouched.

### Docs
- **`docs/algo/opsd.md`** (~1090 lines): background vs OPD, method/equations, both weight-source modes, config reference, example scripts, and the smoke/test plan.

## Examples run

Derived from `examples/on_policy_distillation_trainer/run_qwen3_8b_fsdp.sh`, with the teacher pointed at the student base and the privileged-context flags added:

**Frozen (paper default) — Qwen3-8B student + frozen `θ₀` teacher, GRPO on a math dataset:**
```bash
python3 -m verl.trainer.main_ppo \
  algorithm.adv_estimator=grpo \
  data.train_files="['$HOME/data/math/train.parquet']" \
  data.val_files="['$HOME/data/math/test.parquet']" \
  data.train_batch_size=128 data.max_prompt_length=1024 data.max_response_length=1024 \
  actor_rollout_ref.model.path=Qwen/Qwen3-8B \
  actor_rollout_ref.rollout.name=vllm actor_rollout_ref.rollout.n=1 \
  actor_rollout_ref.actor.optim.lr=1e-6 \
  trainer.n_gpus_per_node=8 trainer.nnodes=1 \
  distillation.enabled=True \
  distillation.n_gpus_per_node=2 distillation.nnodes=1 \
  distillation.self_distillation.enabled=True \
  distillation.self_distillation.teacher_weights=frozen \
  distillation.self_distillation.reference_key=reward_model.ground_truth \
  distillation.self_distillation.max_reference_length=2048 \
  distillation.self_distillation.max_bridge_length=64 \
  '+distillation.teacher_models.self_frozen.key=self_frozen' \
  '+distillation.teacher_models.self_frozen.model_path=Qwen/Qwen3-8B' \
  '+distillation.teacher_models.self_frozen.num_replicas=1' \
  distillation.distillation_loss.loss_mode=forward_kl_topk \
  distillation.distillation_loss.topk=64 \
  distillation.distillation_loss.use_policy_gradient=False \
  distillation.distillation_loss.loss_max_clamp=10.0 \
  distillation.distillation_loss.log_prob_min_clamp=-10.0 \
  distillation.distillation_loss.clip_tau=5.0
```
the log of the experiment is as below
3e57096d094df71ca518ec2e9ed912c4
The eval score as compared with the original paper is as below
image

Credit to @XiangZheng2002

### CODE REVIEW

File: verl/workers/engine/fsdp/transformer_impl.py [hunk 1]
```diff
@@ -1159,7 +1169,15 @@ def prepare_model_outputs(self, output, output_args, micro_batch: TensorDict, lo
 
                 # logits_processor_func return tensors with shape (1, total_nnz/sp_size)
                 if distillation_use_topk:
-                    outputs = logits_processor_func(student_logits=logits_rmpad.unsqueeze(0), data=micro_batch)
+                    # logits_rmpad was divided in-place by temperature_rmpad above (for the policy
+                    # log-probs). Pass that factor so the distillation can undo it and build its
+                    # target from RAW logits (reference OPSD parity; see compute_forward_kl_topk).
+                    student_logits_temperature = temperature_rmpad.clamp(min=1e-8).view(1, -1, 1)
```
[the remark this answers is not in the document]
Comment (XiangZheng2002) [reply verl/workers/engine/fsdp/transformer_impl.py:1175]: > The squeeze is unconditional: `temperature_rmpad = temperature_rmpad.squeeze(0)` sits *outside* the `if self.use_ulysses_sp:` block (same indent level, right after it), so the tensor entering `prepare_model_outputs` is always 1D `(total_nnz,)` regardless of SP. The in-place `div_` at line 1146 then broadcasts `(total_nnz, 1)` against `(total_nnz, vocab)`, which is valid. This unconditional squeeze predates this PR (introduced in #4787).
>
> Empirically verified on both fronts:
> 1. A real OPSD training run with exactly this configuration (`ulysses_sequence_parallel_size=1`, `use_remove_padding=True`, `use_fused_kernels=False`, `loss_mode=forward_kl_topk`, per-sample temperature 1.1) completed 100 steps on 4×A100 without error — the flagged lines executed on every micro-batch.
> 2. A minimal repro driving the real `prepare_model_inputs`/`prepare_model_outputs` with SP disabled confirms `temperature_rmpad` is 1D `(total_nnz,)` and both the `div_` (line 1146) and the top-k distillation path (line 1182) run cleanly; forcing the hypothesized 2D shape does reproduce the predicted RuntimeError, confirming the code path was exercised.
>
> No change needed.

**Comment by XiangZheng2002** [comment]
I wanted to follow up on one issue regarding the "live" and "frozen" modes for the teacher weights.

**On the teacher's weights: frozen vs. live.** The OPSD paper ("Self-Distilled Reasoner", arXiv:2601.18734) is internally inconsistent about *which* weights serve the teacher, and this PR deliberately follows its experiments rather than its main-text definition:

- The **method section (§3, "Teacher and student policies")** defines both policies off the *same, currently-updating* parameters θ — "Both policies share the same parameters θ but differ only in their conditioning context," i.e. `p_T(·|x,y*) ≜ p_θ(·|x,y*)` and `p_S(·|x) ≜ p_θ(·|x)`. This is the "live" teacher, where the teacher tracks the student's current weights.
- The **experiments (§4.1, Implementation details)** instead *freeze* the teacher at the initial checkpoint θ₀: "We fix the teacher policy to be the initial policy, rather than the currently updating learning policy, as we find this helps stabilize training and implicitly acts as regularization to prevent excessive deviation from the initial policy." Every reported number uses this frozen teacher.

This PR implements only the **frozen** teacher (θ₀), which matches the paper's experiments and the reference implementation (`siyan-zhao/OPSD`): reuse the OPD teacher pool with `teacher_models.*.model_path` set to the student's base checkpoint. The **live** teacher (θ served by the rollout engine at current weights) is intentionally out of scope here — it requires additional trainer-level control flow — and will be added in a follow-up PR. Until then, `teacher_weights="live"` fails closed with a `NotImplementedError` so we never silently optimize a different objective.

**Comment by XiangZheng2002** [comment]
Hi — following up with the training-time logs for the three OPSD runs: **verl-opsd** (our
reproduction), **reference-opsd-top256**, and **reference-opsd-full-vocab** (the two official TRL
runs). All three share the same setup — Qwen3-1.7B + LoRA (r64/α128), dataset
`siyanzhao/Openthoughts_math_30k_opsd`, 100 steps, lr 5e-6 linear decay, effective batch 32,
`clip_tau=0.05`, teacher temp 1.0 / student temp 1.1 — and differ only in framework (verl vs TRL)
and the distillation vocab (top-256 vs full).

Since OPSD is pure forward-KL distillation (β=0), the only signal that matters is the
**distillation loss**, which legitimately goes negative once the clip term dominates
(`critic/score/mean` is a constant −1 placeholder). The tables below track it end-to-end.
## Distillation loss

Per-step OPSD distillation loss for all three runs — the single number that tells us whether the
reproduction matches. verl step = optimizer step; the references log every 2 steps at the matching
`global_step`.

| step | **verl** `gqo542a6` | ref top-256 `yprq216h` | ref full-vocab `97rgygv3` |
|-----:|--------------------:|-----------------------:|--------------------------:|
|   0/2 |  0.00655 |  0.00760 |  0.00960 |
|  10/12 |  0.00142 |  0.00120 |  0.00280 |
|  20/22 | −0.00302 | −0.00180 | −0.00100 |
|  30/32 | −0.00496 | −0.00350 | −0.00340 |
|  40/42 | −0.00519 | −0.00480 | −0.00390 |
|  50/52 | −0.00688 | −0.00650 | −0.00640 |
|  60/62 | −0.00830 | −0.00760 | −0.00700 |
|  70/72 | −0.00985 | −0.00850 | −0.00760 |
|  80/82 | −0.01059 | −0.00970 | −0.00900 |
|  90/92 | −0.00931 | −0.00890 | −0.00830 |
| ~100 | **−0.01069** | **−0.01000** | **−0.00970** |

verl tracks both references throughout: same start (~+0.007), crosses zero near step 20, ends at
−0.0107 ≈ reference −0.0100 (top-256) / −0.0097 (full vocab) — equal within logging noise.

**Comment by XiangZheng2002** [comment]
## Per-step × multi-dataset eval matrix — verl vs reference
Beyond the training curves, we also ran the full downstream eval to confirm the two runs behave the
same on actual benchmarks. This sweep evals **every checkpoint (25/50/75/100)** across **3 datasets**
(aime24, aime25, hmmt25) under the paper's Table-2 protocol (`enable_thinking=True`,
`temperature=1.0`, `top_k=-1`, `max_new_tokens=38912`, `val_n=12` → 30 problems × 12 =
**360 samples/cell**). Both sides are **top-256 OPSD** (the eval's `--top_k -1` is unrestricted
*sampling* at generation time, unrelated to the training-time top-256 distillation truncation).


### Avg@12 (main metric)

The headline eval: mean accuracy over 12 samples/problem (%), the metric OPSD is judged on. Rows =
training step; columns = dataset × framework (`v` = verl, `r` = reference).

| step | aime24 v | aime24 r | aime25 v | aime25 r | hmmt25 v | hmmt25 r |
|-----:|--------:|--------:|--------:|--------:|--------:|--------:|
| base | 46.67 | 46.67 | 36.94 | 36.94 | 22.50 | 22.50 |
| 25 | 53.61 | 51.11 | 39.17 | 38.89 | 25.28 | 21.67 |
| 50 | 54.44 | 55.56 | 38.33 | 35.56 | 24.72 | 27.78 |
| 75 | 53.89 | 57.22 | 43.33 | 41.11 | 25.00 | 27.22 |
| **100** | **58.33** | **54.44** | **41.11** | **41.67** | **27.50** | **26.39** |
| **Δ (100−base)** | **+11.66** | **+7.77** | **+4.17** | **+4.73** | **+5.00** | **+3.89** |

`base` (step 0) is the shared starting checkpoint, so its `v`/`r` columns are identical.

### Pass@12 / Maj@12

Two complementary views of the same runs: **Pass@12** = fraction of problems solved by ≥1 of the 12
samples (coverage / diversity), **Maj@12** = fraction correct under majority vote (single-answer
reliability). Together they show the exploration↔exploitation trade-off distillation induces —
Avg@12 and Maj@12 rise while Pass@12 can dip slightly on verl. Same layout as above.

**Pass@12**

| step | aime24 v | aime24 r | aime25 v | aime25 r | hmmt25 v | hmmt25 r |
|-----:|--------:|--------:|--------:|--------:|--------:|--------:|
| base | 76.67 | 76.67 | 73.33 | 73.33 | 46.67 | 46.67 |
| 25 | 80.00 | 76.67 | 63.33 | 66.67 | 50.00 | 43.33 |
| 50 | 76.67 | 76.67 | 60.00 | 60.00 | 56.67 | 53.33 |
| 75 | 80.00 | 76.67 | 70.00 | 63.33 | 46.67 | 56.67 |
| **100** | **76.67** | **76.67** | **66.67** | **73.33** | **43.33** | **46.67** |
| **Δ (100−base)** | **0.00** | **0.00** | **−6.66** | **0.00** | **−3.34** | **0.00** |

**Maj@12**

| step | aime24 v | aime24 r | aime25 v | aime25 r | hmmt25 v | hmmt25 r |
|-----:|--------:|--------:|--------:|--------:|--------:|--------:|
| base | 63.33 | 63.33 | 43.33 | 43.33 | 23.33 | 23.33 |
| 25 | 70.00 | 66.67 | 53.33 | 53.33 | 30.00 | 23.33 |
| 50 | 63.33 | 70.00 | 50.00 | 46.67 | 26.67 | 30.00 |
| 75 | 66.67 | 70.00 | 60.00 | 50.00 | 26.67 | 30.00 |
| **100** | **73.33** | **70.00** | **50.00** | **56.67** | **36.67** | **30.00** |
| **Δ (100−base)** | **+10.00** | **+6.67** | **+6.67** | **+13.34** | **+13.34** | **+6.67** |

`base` is the shared starting checkpoint, so its `v`/`r` columns are identical.

---

image


*Per-step eval curves for all three tables above, as a 3×3 grid: **rows = metric** (Avg@12 = mean accuracy over 12 samples; Pass@12 = solved by ≥1 of 12 samples; Maj@12 = correct under majority vote), **columns = dataset** (aime24 / aime25 / hmmt25). Each panel plots accuracy (%) vs training step (base=0, 25, 50, 75, 100); **solid blue = verl**, **dashed orange = reference**. Per-panel y-axes auto-scale independently (the three datasets sit at very different accuracy levels), so read each panel for the verl-vs-reference gap and trend, not for cross-panel height.*
